src/Data/HashMap/Strict/InsOrd.hs

```diff
@@ -400,8 +412,65 @@ mapWithKey f (InsOrdHashMap i m) =
   where
     f' k (P j x) = P j (f k x)
 
+foldMapWithKey :: Monoid m => (k -> a -> m) -> InsOrdHashMap k a -> m
+foldMapWithKey f = foldMap (uncurry f) . toList
+
 traverseWithKey :: Applicative f => (k -> a -> f b) -> InsOrdHashMap k a -> f (InsOrdHashMap k b)
-traverseWithKey f (InsOrdHashMap i m) =
+traverseWithKey f (InsOrdHashMap n m) = InsOrdHashMap n <$> retractSortedAp
+    (HashMap.traverseWithKey (\k (P i v) -> liftSortedAp i (P i <$> f k v)) m)
+
+-- Sort using insertion sort
+-- Hopefully it's fast enough for where we need it
+-- otherwise: https://gist.github.com/treeowl/9621f58d55fe0c4f9162be0e074b1b29
+-- http://elvishjerricco.github.io/2017/03/23/applicative-sorting.html also related
+
+-- Free applicative which re-orders effects
+-- Mostly from Edward Kmett's `free` package.
```

Comment:
`free` is a transitive dependency of `insert-ordered-containers`. We might use them directly! Except there's no `SortedAp` defined there at this time. Maybe we should add one there?

Also, `SortedAp f` seems to be like `Ap (Compose P f)` except with sorting built into `<*>`.
It looks like we can use just `Ap` together with something like:

```haskell
sortAp :: Ord1 f => Ap f a -> Ap f a
sortApWith :: (forall a. f a -> f a -> Ordering) -> Ap f b -> Ap f b

-- this instance is not ideal
instance Ord1 (Compose P f) where
  compare1 (Compose (P i _)) (Compose (P j _)) = compare1 i j
```